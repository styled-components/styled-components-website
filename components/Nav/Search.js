import { Search as SearchIcon } from '@styled-icons/material';
import PropTypes from 'prop-types';
import React, { useEffect, useRef } from 'react';
import styled, { createGlobalStyle, css } from 'styled-components';
import { blmGrey, blmLightGrey, darkGrey, grey } from '../../utils/colors';
import { resetInput } from '../../utils/form';
import { mobile } from '../../utils/media';
import rem from '../../utils/rem';
import { navbarHeight } from '../../utils/sizes';

const StyledSearchIcon = styled(SearchIcon)``;

const INPUT_ID = 'docs-search-input';

const Wrapper = styled.form`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  flex: 0 0 auto;
  ${mobile(css`
    display: block;
    padding: 0;
    span.algolia-autocomplete {
      display: block !important;
    }
    span.ds-dropdown-menu::before {
      display: none;
    }
  `)};
`;

const Input = styled.input`
  ${resetInput};
  appearance: none;
  flex: 0 0 auto;
  width: ${rem(130)};
  line-height: ${rem(navbarHeight - 20)};
  font-size: ${rem(15)};
  color: currentColor;

  ::placeholder {
    color: currentColor;
    opacity: 0.5;
    transition: opacity 0.2s ease-in-out;
  }

  :focus {
    ::placeholder {
      opacity: 0.8;
    }
  }

  ${mobile(css`
    padding: ${rem(10)} ${rem(48)};
    display: block;
    width: 100%;
    background: ${blmGrey};
    color: white;
  `)};
`;

const Button = styled.label.attrs((/* props */) => ({
  htmlFor: INPUT_ID,
}))`
  ${resetInput};
  flex: 0 0 auto;
  ${resetInput} flex: 0 0 auto;
  margin-right: ${rem(4)};
  cursor: pointer;
  display: flex;
  &:hover,
  &:focus {
    opacity: 0.7;
    filter: drop-shadow(0 1px 2px rgba(0, 0, 0, 0.2));
  }

  &:active {
    background: rgba(0, 0, 0, 0.1);
  }

  ${StyledSearchIcon} {
    width: ${rem(26)};
    height: ${rem(26)};
  }

  ${mobile(css`
    position: absolute;
    top: ${rem(9)};
    left: ${rem(10)};
    height: ${rem(32)};
    display: flex;
    align-items: center;
    justify-content: center;
    width: ${rem(32)};
    z-index: 1;
    background: none;
    color: white;
  `)};
`;

const GlobalStyles = createGlobalStyle`
  .algolia-autocomplete
  .ds-dropdown-menu
  .ds-suggestion
  .algolia-docsearch-suggestion--content {
    width: 100% !important;

    &::before {
      content: none;
    }
  }

  .algolia-autocomplete
    .ds-dropdown-menu
    .ds-suggestion.ds-cursor
    .algolia-docsearch-suggestion--content {
    background: ${blmLightGrey} !important;
  }

  /* Main category (eg. Getting Started) */
  .algolia-autocomplete .algolia-docsearch-suggestion--category-header {
    color: ${darkGrey};
  }

  /* Category (eg. Downloads) */
  .algolia-autocomplete .algolia-docsearch-suggestion--subcategory-column {
    display: none !important;
  }

  /* Title (eg. Bootstrap CDN) */
  .algolia-autocomplete .algolia-docsearch-suggestion--title {
    font-weight: bold;
    color: black;
  }

  /* Description description (eg. Bootstrap currently works...) */
  .algolia-autocomplete .algolia-docsearch-suggestion--text {
    color: ${grey};
  }

  /* Highlighted text */
  .algolia-autocomplete .algolia-docsearch-suggestion--highlight {
    box-shadow: none !important;
    color: ${blmGrey} !important;
    background: transparent !important;
  }

  .algolia-autocomplete .ds-dropdown-menu {
    margin-top: 0;

    > *:first-child {
      border: 0;
    }
  }
`;

const Search = ({ isDocs, className }) => {
  const searchInput = useRef(null);

  useEffect(() => {
    if (searchInput.current) searchInput.current.focus();
  }, []);

  return (
    <Wrapper className={className}>
      <GlobalStyles />
      <Button>
        <StyledSearchIcon />
      </Button>
      <Input id={INPUT_ID} placeholder={isDocs ? `Search ...` : `Search docs ...`} ref={searchInput} type="search" />
    </Wrapper>
  );
};

Search.propTypes = {
  className: PropTypes.string,
};

export default Search;
