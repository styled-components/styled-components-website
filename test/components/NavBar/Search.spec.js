import React from 'react';
import renderer from 'react-test-renderer';
import Search from '../../../components/Nav/Search';

test('Search renders correctly', () => {
  const tree = renderer.create(<Search />).toJSON();

  expect(tree).toMatchSnapshot();
});
