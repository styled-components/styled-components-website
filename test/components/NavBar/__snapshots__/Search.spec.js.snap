// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c2 {
  background: none;
  outline: none;
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 7.222222222222222rem;
  line-height: 1.6666666666666667rem;
  font-size: 0.8333333333333334rem;
  color: currentColor;
}

.c2::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.5;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
}

.c2::-moz-placeholder {
  color: currentColor;
  opacity: 0.5;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
}

.c2:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.5;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
}

.c2::placeholder {
  color: currentColor;
  opacity: 0.5;
  -webkit-transition: opacity 0.2s ease-in-out;
  transition: opacity 0.2s ease-in-out;
}

.c2:focus::-webkit-input-placeholder {
  opacity: 0.8;
}

.c2:focus::-moz-placeholder {
  opacity: 0.8;
}

.c2:focus:-ms-input-placeholder {
  opacity: 0.8;
}

.c2:focus::placeholder {
  opacity: 0.8;
}

.c1 {
  background: none;
  outline: none;
  border: none;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  background: none;
  outline: none;
  border: none;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin-right: 0.2222222222222222rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1:hover,
.c1:focus {
  opacity: 0.7;
  -webkit-filter: drop-shadow(0 1px 2px rgba(0,0,0,0.2));
  filter: drop-shadow(0 1px 2px rgba(0,0,0,0.2));
}

.c1:active {
  background: rgba(0,0,0,0.1);
}

.c1 .{
  width: 1.4444444444444444rem;
  height: 1.4444444444444444rem;
}

@media (max-width:62.5em) {
  .c0 {
    display: block;
    padding: 0;
  }

  .c0 span.algolia-autocomplete {
    display: block !important;
  }

  .c0 span.ds-dropdown-menu::before {
    display: none;
  }
}

@media (max-width:62.5em) {
  .c2 {
    padding: 0.5555555555555556rem 2.6666666666666665rem;
    display: block;
    width: 100%;
    background: rgb(33,33,33);
    color: white;
  }
}

@media (max-width:62.5em) {
  .c1 {
    position: absolute;
    top: 0.5rem;
    left: 0.5555555555555556rem;
    height: 1.7777777777777777rem;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    width: 1.7777777777777777rem;
    z-index: 1;
    background: none;
    color: white;
  }
}

<form
  className="c0"
>
  <label
    className="c1"
    htmlFor="docs-search-input"
  >
    <svg
      className=""
    />
  </label>
  <input
    className="c2"
    id="docs-search-input"
    placeholder="Search docs ..."
    type="search"
  />
</form>
`;
